### PR TITLE
[FLINK-25072][streaming] Introduce description on Transformation and …

### PR DESCRIPTION
**Pull request by wenlong88**
…StreamGraph

## What is the purpose of the change
Introduce an optional field `description` on transformation and StreamGraph which allow use to set a more detailed information about an operation without affecting name of job vertex, will be used to construct description of job vertex in FLINK-25073.

## Brief change log

  - *introduce a new interface to set description on DataStream API*
  - *transpose the description from Transformation to StreamGraph*

## Verifying this change

This change added tests and can be verified as follows:

*(example:)*
  - *Added tests in DataStreamTest to make sure that description is transposed exactly *

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

**Review by gaoyunhaii — COMMENTED** [review]
Hi @wenlong88 very thanks for opening the PR! May I confirm that does this PR intend to modify all the related UI? If so, we might also need to check the related rest API to see if they are all modified.

**Comment by wenlong88** [comment]
@gaoyunhaii  thanks for the review, this pr just adds the description in transformation so that I can make the pr small and clear, and create pr on DS and SQL concurrently once this pr is merged. regarding UI changes, it is in #17938. as you can see in #17938 , we will not change the format of response in rest api, we just change the content of JSON plan in rest api and the way to display in UI.

**Comment by wenlong88** [comment]
@flinkbot run azure

**Comment by gaoyunhaii** [comment]
LGTM from my side. And @AHeise do you want to also have a look~? Very thanks!

**Comment by wenlong88** [comment]
@AHeise could you help check if there are any other comments on the pr? It would be great if we can finish the review ASAP, because it is quite simple IMO and we have some other subtasks waiting this pr.

File: flink-core/src/main/java/org/apache/flink/api/dag/Transformation.java
Comment (AHeise) [reply flink-core/src/main/java/org/apache/flink/api/dag/Transformation.java]: Please do to also keep getter and field in sync.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/SingleOutputStreamOperator.java
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/SingleOutputStreamOperator.java:434]: Should this be directly exposed on `DataStream`? For which operator wouldn't you want it to be set.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamNode.java
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamNode.java]: Please do.

File: flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/DataStream.scala
Comment (AHeise) [inline flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/DataStream.scala:1240]: If we don't really need it in DataStream (see top-level comment), please remove.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamNode.java
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamNode.java]: I'd move this logic to call-site. Either while creating the StreamNode or while reading the StreamNode.

**Review by AHeise — CHANGES_REQUESTED** [review]
I have not fully understood the scope of the PR. Is this relevant for `DataStream` users or only for Table/SQL?
Maybe a different name as description would be better. Is this description also used for rendering in Flink Web UI? Currently, you mentioned only Json plan. I don't think most users know exactly what that is and where it's used.

If also for `DataStream`: When would a user set it? What is the relation between name, uid, and description? What happens if the user sets description but not the name? I think the addition makes it harder to reason about the expected results. Please expand the JavaDoc to answer these question if you think there is value to a `DataStream` user. 

If only for Table/SQL: Please remove the additions to `DataStream`.

Note that a quick solution would be to remove the additions to `DataStream` for now to unblock Table/SQL and add a second PR where we discuss the user-facing things. Transformation/StreamNode is internal, so all changes are easier to make and reverse.

**Comment by wenlong88** [comment]
hi, @AHeise  thanks for the review. The description would be used in both DataStream and SQL/Table API according to the FLIP discussion, would be used in JSON plan and finally affects content in Web UI, the related change is in [#17938](https://github.com/apache/flink/pull/17938). Improvement on DataStream operators will be in following up issues such as https://issues.apache.org/jira/browse/FLINK-25074.

About the naming and whether we need to expose API in DataStream, I think we have reached a consensus in the FLIP discussion, We are trying to introduce descriptions so that we can simplify the names. This is worth in both SQL and DataStream API IMO. Do you have any more suggestion or comments about this? It would be great if there is some better idea.

I have updated java docs and addressed other comments, please check again, thanks.

**Comment by wenlong88** [comment]
hi, @AHeise I also am ok if you really think it is necessary to remove the api changes currently, and involve more discussion. I have rebase the #17938 with the API change removed, you can checkout there to see how we are going to use the description first.

**Review by AHeise — COMMENTED** [review]
Sorry I haven't seen that this went through a whole FLIP process already. It would have been nice to cross-reference that in the PR description but it's indirectly referenced through the parent ticket.

In any case, I'd retract my concerns as this has been publicly discussed and accepted. So please address the detailed feedback and we are good to go.

The JavaDoc (and markdown) docs should capture where name/description is used exactly. I think you did a good job with your last commits in the JavaDoc. I'm not sure if there is an obvious place in the `docs` where you'd also need to add it.

**Comment by wenlong88** [comment]
hi, @AHeise I didn't find somewhere about the introduction of api such as uid and name while adding the description, I think it may be because most of the names/descriptions in DataStream API are generated by flink, and for advanced users java doc maybe enough when well described?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamNode.java
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamNode.java:117]: How about also providing `operatorDescription` as a constructor parameter? It doesn't look that different from `operatorName`. Then you can make the field `final` and get rid of the setter. While creating the streamNode, you would then simply either pass `transformation.getOperatorDescription()` or `operatorName` if the former is null.

**Review by AHeise — COMMENTED** [review]
Last round of smaller nits and suggestions.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamNode.java
Comment (wenlong88) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamNode.java:117]: Using a separated setter can avoid touching all of the constructor of StreamNode, which could be quite a lot of code including helper method in StreamGraph and all of the TransformationTranslaters, and setting property of StreamNode in SimpleTransformationTranslator#configure is widely used. so I prefer init it by a setter in configure.
what do you think?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/DataStreamSink.java
Comment (wenlong88) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/DataStreamSink.java:157]: check added in Transformation

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamNode.java
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamNode.java:117]: Fair point. Let's go with that.

**Review by AHeise — APPROVED** [review]
Thank you very much for your contribution. LGTM.

Regarding adjusting the markdown docs, I'd propose to either add it to https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/dev/datastream/overview/#debugging or add a dedicated section to 
https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/ops/debugging/ . It makes most sense to describe it once the feature is fully merged.

And one safety reminder: if you cannot merge the feature fully to 1.15 for some reason, please consider hiding the DataStream portions of this PR to not confuse users.

**Comment by wenlong88** [comment]
thanks again for the review, @AHeise , I have created a issue for further completing the doc: https://issues.apache.org/jira/browse/FLINK-25198

**Comment by gaoyunhaii** [comment]
Very thanks @AHeise for the detailed review! Merging~

**Pull request closed without merging**
